cont3xt/integration.js

```diff
@@ -379,25 +379,40 @@ class Integration {
     return res.send({ success: true, integrations: results });
   }
 
-  static async runIntegrationsList (shared, query, itype, integrations, onFinish) {
+  static async runIntegrationsList (shared, indicator, parentQuery, integrations, onFinish) {
+    const { query, itype } = indicator;
+
+    // initial write (ensures dependable tracking of indicator tree)
+    shared.res.write(JSON.stringify({ purpose: 'link', indicator, parentQuery }));
+    shared.res.write(',\n');
+
+    // do not reissue integrations if they have been started by a matching query reached via different descendants
+    if (shared.queriedSet.has(query)) { return; }
+    shared.queriedSet.add(query);
+
+    // update integration total
     shared.total += integrations.length;
 
     if (Integration.debug > 0) {
       console.log('RUNNING', itype, query, integrations.map(integration => integration.name));
     }
 
     const writeOne = (integration, response) => {
-      if (integration.addMoreIntegrations) {
-        integration.addMoreIntegrations(itype, response, (moreQuery, moreIType) => {
-          Integration.runIntegrationsList(shared, moreQuery, moreIType, Integration.#integrations[moreIType], onFinish);
+      if (integration.addMoreIntegrations) { // TODO: make addMoreIntegrations take indicator
+        integration.addMoreIntegrations(itype, response, (moreIndicator, enhanceInfo = undefined) => {
+          if (enhanceInfo != null && typeof enhanceInfo === 'object') {
+            shared.res.write(JSON.stringify({ purpose: 'enhance', indicator: moreIndicator, enhanceInfo }));
+            shared.res.write(',\n');
+          }
+          Integration.runIntegrationsList(shared, moreIndicator, query, Integration.#integrations[moreIndicator.itype], onFinish);
```

Comment:
In this specific case, no, since the enhanceInfo from addMore is intended for moreIndicator (so I guess I should probably add a null check for that?). If what you mean is can we use this for things other than moreIndicator, yes—so maybe it should be extracted into a function? It's not used anywhere else though, and I'm not sure what even would cause an enhancement that isn't an integration